pkg/gardenlet/controller/managedseed/managedseed_actuator.go

```diff
@@ -0,0 +1,835 @@
+	// Get seed object
+	seed := &gardencorev1beta1.Seed{}
+	if err := a.gardenClient.Client().Get(ctx, kutil.Key(managedSeed.Name), seed); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	// Delete seed object
+	if err := a.gardenClient.Client().Delete(ctx, seed); client.IgnoreNotFound(err) != nil {
+		return err
+	}
+
+	// Check if there are any remaining objects associated with this seed
+	if err := checkSeedAssociations(ctx, a.gardenClient.Client(), seed.Name); err != nil {
+		return err
+	}
+
+	// Delete seed secrets
+	if seed.Spec.SecretRef != nil {
+		if err := kutil.DeleteSecretByRef(ctx, a.gardenClient.Client(), *seed.Spec.SecretRef); err != nil {
+			return err
+		}
+	}
+	if seed.Spec.Backup != nil {
+		if err := kutil.DeleteSecretByRef(ctx, a.gardenClient.Client(), seed.Spec.Backup.SecretRef); err != nil {
+			return err
+		}
+	}
+
+	// Return an error since the seed still exists
+	return fmt.Errorf("seed %q still exists", seed.Name)
+}
+
+func (a *actuator) deployGardenlet(ctx context.Context, managedSeed *seedmanagementv1alpha1.ManagedSeed, shoot *gardencorev1beta1.Shoot) error {
+	// Get gardenlet chart values
+	values, err := a.getGardenletChartValues(ctx, managedSeed, shoot)
+	if err != nil {
+		return err
+	}
+
+	// Apply gardenlet chart
+	return a.shootClient.ChartApplier().Apply(ctx, filepath.Join(common.ChartPath, "gardener", "gardenlet"), v1beta1constants.GardenNamespace, "gardenlet", kubernetes.Values(values))
+}
+
+func (a *actuator) deleteGardenlet(ctx context.Context, managedSeed *seedmanagementv1alpha1.ManagedSeed, shoot *gardencorev1beta1.Shoot) error {
+	// Get gardenlet chart values
+	values, err := a.getGardenletChartValues(ctx, managedSeed, shoot)
+	if err != nil {
+		return err
+	}
+
+	// Delete gardenlet chart
+	if err := a.shootClient.ChartApplier().Delete(ctx, filepath.Join(common.ChartPath, "gardener", "gardenlet"), v1beta1constants.GardenNamespace, "gardenlet", kubernetes.Values(values)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (a *actuator) prepareSeedSpec(ctx context.Context, managedSeed *seedmanagementv1alpha1.ManagedSeed, shoot *gardencorev1beta1.Shoot, seedSpec *gardencorev1beta1.SeedSpec) (*gardencorev1beta1.SeedSpec, error) {
+	// Get shoot secret
+	shootSecret, err := a.getShootSecret(ctx, shoot)
+	if err != nil {
+		return nil, err
+	}
+
+	// Prepare seed backup
+	seedBackup, err := a.prepareSeedBackup(ctx, managedSeed, shoot, shootSecret, seedSpec.Backup)
+	if err != nil {
+		return nil, err
+	}
+
+	// Copy seed spec
+	spec := seedSpec.DeepCopy()
+
+	// Initialize backup
+	spec.Backup = seedBackup
+
+	// Initialize DNS
+	if spec.DNS.IngressDomain == "" && shoot.Spec.DNS != nil && shoot.Spec.DNS.Domain != nil {
+		spec.DNS.IngressDomain = fmt.Sprintf("%s.%s", common.IngressPrefix, *shoot.Spec.DNS.Domain)
+	}
```

Comment:
I will take a look. I guess it's a new feature since the Gardener this PR is based upon doesn't even have `Ingress` in the seed spec. BTW, the code above is copied from the current `use-as-seed` implementation.

Reply:
I adapted the code recently introduced by @BeckerMax in the original seed registration controller. The managed seed controller should now behave identically if `seed.spec.ingress.controller` is set, except that if `seed.spec.dns.provider` is also set it will be used, instead of being computed from the shoot (values specified by the user take precedence over computed values)